# 04-1 통계로 요약하기

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://nbviewer.jupyter.org/github/rickiepark/hg-da/blob/main/04-1.ipynb"><img src="https://jupyter.org/assets/share.png" width="61" />주피터 노트북 뷰어로 보기</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/rickiepark/hg-da/blob/main/04-1.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />구글 코랩(Colab)에서 실행하기</a>
  </td>
</table>

## 기술통계 구하기

In [1]:
import gdown

gdown.download('https://bit.ly/3736JW1', 'ns_book6.csv', quiet=False)

Downloading...
From: https://bit.ly/3736JW1
To: C:\Users\user\Desktop\Han_Toss\hg-da-main\ns_book6.csv
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 55.0M/55.0M [00:31<00:00, 1.74MB/s]


'ns_book6.csv'

In [2]:
import pandas as pd

ns_book6 = pd.read_csv('ns_book6.csv', low_memory=False)
ns_book6.head()

C:\Users\user\AppData\Local\Temp\ipykernel_8152\550080778.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


번호                    도서명                저자    출판사    발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021.0  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021.0  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021.0  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021.0  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021.0  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19

In [3]:
ns_book6.describe()

번호           발행년도           도서권수           대출건수
count  379976.000000  379976.000000  379976.000000  379976.000000
mean   201726.332847    2008.516306       1.135874      11.504629
std    115836.454596       8.780529       0.483343      19.241926
min         1.000000    1947.000000       0.000000       0.000000
25%    102202.750000    2003.000000       1.000000       2.000000
50%    203179.500000    2009.000000       1.000000       6.000000
75%    301630.250000    2015.000000       1.000000      14.000000
max    401681.000000    2650.000000      40.000000    1765.000000

In [4]:
sum(ns_book6['도서권수']==0)

3206

In [5]:
ns_book7 = ns_book6[ns_book6['도서권수']>0]

In [6]:
ns_book7.describe(percentiles=[0.3, 0.6, 0.9])

번호           발행년도           도서권수           대출건수
count  376770.000000  376770.000000  376770.000000  376770.000000
mean   202977.476649    2008.460076       1.145540      11.593439
std    115298.245784       8.773148       0.473853      19.279409
min         1.000000    1947.000000       1.000000       0.000000
30%    124649.700000    2004.000000       1.000000       2.000000
50%    204550.500000    2009.000000       1.000000       6.000000
60%    243537.400000    2011.000000       1.000000       8.000000
90%    361341.100000    2018.000000       2.000000      28.000000
max    401681.000000    2650.000000      40.000000    1765.000000

In [7]:
ns_book7.describe(include='object')

도서명             저자     출판사           ISBN        세트 ISBN    부가기호  \
count   376770         376770  376770         376770          55866  308252   
unique  336408         248850   21875         350810          14875      17   
top      승정원일기  세종대왕기념사업회 [편]    문학동네  9788937430299  9788937460005       0   
freq       250            303    4410            206            702  158235   

            권  주제분류번호        등록일자  
count   61793  359792      376770  
unique    834   12467        4562  
top         1   813.6  1970-01-01  
freq    13282   14816       28185

## 평균

$평균 = \dfrac{a + b + c}{3}$

$평균 = \dfrac{x_1 + x_2 + x_3}{3}$

In [8]:
x = [10, 20, 30]
sum = 0
for i in range(3):
    sum += x[i]
print("평균:", sum / len(x))

평균: 20.0


$평균 = \dfrac{x_1 + x_2 + x_3}{3} = \dfrac{\sum_{i=1}^{3} x_i}{3}$

$평균 대출건수 = \dfrac{\sum_{i=1}^{376770} x_i}{376770}$

In [9]:
ns_book7['대출건수'].mean()

11.593438968070707

## 중앙값

In [10]:
ns_book7['대출건수'].median()

6.0

In [11]:
temp_df = pd.DataFrame([1,2,3,4])
temp_df.median()

0    2.5
dtype: float64

In [12]:
ns_book7['대출건수'].drop_duplicates().median()

183.0

## 최솟값, 최댓값

In [13]:
ns_book7['대출건수'].min()

0

In [14]:
ns_book7['대출건수'].max()

1765

In [17]:
ns_book7.loc[ns_book7['대출건수'] == ns_book7['대출건수'].max(),:].head(1)

번호                                      도서명                 저자  출판사  \
53114  55565  사피엔스 :유인원에서 사이보그까지, 인간 역사의 대담하고 위대한 질문   유발 하라리 지음 ;조현욱 옮김  김영사   

         발행년도           ISBN 세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
53114  2016.0  9788934972464     NaN  NaN  NaN    909     1  1765  2018-04-03

## 분위수

In [15]:
ns_book7['대출건수'].quantile(0.25)

2.0

In [16]:
ns_book7['대출건수'].quantile([0.25,0.5,0.75])

0.25     2.0
0.50     6.0
0.75    14.0
Name: 대출건수, dtype: float64

In [17]:
pd.Series([1,2,3,4,5]).quantile(0.9)

4.6

In [18]:
4 + (0.9-0.75)*(5-4)/(1.0-0.75)

4.6

In [19]:
pd.Series([1,2,3,4,5]).quantile(0.9, interpolation='midpoint')

4.5

In [20]:
pd.Series([1,2,3,4,5]).quantile(0.9, interpolation='nearest')

5

In [21]:
borrow_10_flag = ns_book7['대출건수'] < 10

In [22]:
borrow_10_flag.mean()

0.6402712530190833

In [23]:
ns_book7['대출건수'].quantile(0.65)

10.0

## 분산

$ s^2 = \dfrac{\sum_{i=1}^{n}(x_i - \bar{x})^2}{n}$

$ \bar{x} = \dfrac{\sum_{i=1}^n x_i}{n}$

In [24]:
ns_book7['대출건수'].var()

371.69563042906674

## 표준 편차

$ s = \sqrt{\dfrac{\sum_{i=1}^{n}(x_i - \bar{x})^2}{n}}$

In [25]:
ns_book7['대출건수'].std()

19.279409493785508

$\sqrt{4}=2$

In [26]:
import numpy as np

diff = ns_book7['대출건수'] - ns_book7['대출건수'].mean()

np.sqrt( np.sum(diff**2) / (len(ns_book7)-1) )

19.279409493785508

## 최빈값

In [27]:
ns_book7['도서명'].mode()

0    승정원일기
Name: 도서명, dtype: object

In [28]:
ns_book7['발행년도'].mode()

0    2012.0
Name: 발행년도, dtype: float64

## 데이터프레임에서 기술통계 구하기

In [29]:
ns_book7.mean(numeric_only=True)

번호      202977.476649
발행년도      2008.460076
도서권수         1.145540
대출건수        11.593439
dtype: float64

In [30]:
ns_book7.loc[:, '도서명':].mode()

도서명             저자   출판사    발행년도           ISBN        세트 ISBN 부가기호  권  \
0  승정원일기  세종대왕기념사업회 [편]  문학동네  2012.0  9788937430299  9788937460005    0  1   

  주제분류번호  도서권수  대출건수        등록일자  
0  813.6     1     0  1970-01-01

In [31]:
ns_book7.to_csv('ns_book7.csv', index=False)

## 넘파이의 기술통계 함수

### 평균 구하기

In [32]:
import numpy as np

np.mean(ns_book7['대출건수'])

11.593438968070707

$\dfrac{국어점수 \times 2 + 수학점수}{3}$

$\dfrac{국어점수 \times 국어가중치 + 수학점수 \times 수학가중치}{국어가중치 + 수학가중치}$

$가중평균 = \dfrac{x_1 \times w_1 + x_2 \times w_2}{w_1 + w_2} = \dfrac{\sum_{i=1}^{2} x_i \times w_i}{\sum_{i=1}^{2} w_i}$

In [33]:
np.average(ns_book7['대출건수'], weights=1/ns_book7['도서권수'])

10.543612175385386

In [34]:
np.mean(ns_book7['대출건수']/ns_book7['도서권수'])

9.873029861445774

In [35]:
ns_book7['대출건수'].sum()/ns_book7['도서권수'].sum()

10.120503701300958

### 중앙값 구하기

In [36]:
np.median(ns_book7['대출건수'])

6.0

### 최솟값, 최댓값 구하기

In [37]:
np.min(ns_book7['대출건수'])

0

In [38]:
np.max(ns_book7['대출건수'])

1765

### 분위수 구하기

In [39]:
# interpolation 매개변수가 numpy 1.22(python >= 3.8) 버전부터 method로 바뀜
np.quantile(ns_book7['대출건수'], [0.25,0.5,0.75])

array([ 2.,  6., 14.])

### 분산 구하기

In [40]:
np.var(ns_book7['대출건수'])

371.6946438971496

$ s^2 = \dfrac{\sum_{i=1}^{n}(x_i - \bar{x})^2}{n - 1}$

In [41]:
ns_book7['대출건수'].var(ddof=0)

371.6946438971496

In [42]:
np.var(ns_book7['대출건수'], ddof=1)

371.69563042906674

### 표준 편차 구하기

In [43]:
np.std(ns_book7['대출건수'])

19.27938390865096

### 최빈값 구하기

In [44]:
values, counts = np.unique(ns_book7['도서명'], return_counts=True)
max_idx = np.argmax(counts)
values[max_idx]

'승정원일기'

## 확인문제

#### 4.

In [45]:
ns_book7[['출판사','대출건수']].groupby('출판사').mean().sort_values('대출건수', ascending=False).head(10)

대출건수
출판사                        
동녁라이프            224.000000
와이비엠시사           215.000000
환상               210.000000
두앤비컨텐츠(랜덤하우스중앙)  167.000000
개미들출판사-에셀문화사     157.000000
해커스 어학 연구소       149.500000
Test clinic      137.000000
마야출판사            133.333333
해밀 & Co.         133.000000
마그나(Magna)       130.000000

#### 5.

In [46]:
target_range = np.array(ns_book7['대출건수'].quantile(q=[0.25,0.75]))
target_bool_idx = (ns_book7['대출건수'] >= target_range[0]) & (ns_book7['대출건수'] <= target_range[1])
target_bool_idx.sum()/len(ns_book7)*100

51.51737134060568